src/odbc_connection.cpp

```diff
@@ -2844,108 +2913,399 @@ fetch_and_store
+                      (
+                        row[column_index].wchar_data,
+                        row[column_index].size + data_to_copy_length
+                      );
+                      memcpy
+                      (
+                        row[column_index].wchar_data + row[column_index].size,
+                        data->long_data_buffer,
+                        data_to_copy_length
+                      );
+                    case SQL_C_CHAR:
+                    default:
+                      data_to_copy_length = strlen(data->long_data_buffer);
+                      row[column_index].char_data =
+                      (SQLCHAR *)
+                      realloc
+                      (
+                        row[column_index].char_data,
+                        row[column_index].size + data_to_copy_length
+                      );
+                      memcpy
+                      (
+                        row[column_index].char_data + row[column_index].size,
+                        data->long_data_buffer,
+                        data_to_copy_length
+                      );
+                  }
+                  row[column_index].size = row[column_index].size + data_to_copy_length;
+
+                  // Once data has been copied, increase to buffer size to fit
+                  // the full size of the data, hopefully avoiding future
+                  // resizes
+                  data->long_data_buffer_size = data->long_data_buffer_size * 1.5;
+
+                  data->long_data_buffer =
+                  (char *) realloc(
+                    data->long_data_buffer,
+                    data->long_data_buffer_size
                   );
-                  break;
-                }
 
-                case SQL_C_WCHAR:
-                {
-                  SQLWCHAR *memory_start = (SQLWCHAR *)data->bound_columns[column_index].buffer + (row_index * data->columns[column_index]->ColumnSize + 1);
-                  row[column_index].size = strlen16((const char16_t *)memory_start);
-                  row[column_index].wchar_data = new SQLWCHAR[row[column_index].size + 1]();
-                  memcpy
+                  data->sqlReturnCode =
+                  SQLGetData
                   (
-                    row[column_index].wchar_data,
-                    memory_start,
-                    row[column_index].size * sizeof(SQLWCHAR)
+                    data->hSTMT,
+                    column_index + 1,
+                    data->columns[column_index]->bind_type,
+                    data->long_data_buffer,
+                    data->long_data_buffer_size,
+                    &data->long_data_string_length_or_indicator
                   );
-                  break;
                 }
 
-                case SQL_C_CHAR:
-                default:
+                // At last, SQLGetData has told us how much data is in the
+                // buffer that was returned! Move the data one last time,
```

Comment:
SQLGetData doesn't return how much was returned -- it returns how much was returned + how much is left to return.

If you have a 300 byte buffer and there's a 1000 bytes of data the calls could look like this:

- indicator -> 1000
- indicator -> 700
- indicator -> 400
- indicator -> 100

With variable length data, it could also look like this:

- indicator -> 1000 (returned 298 bytes)
- indicator -> 702 (returned 297 byes)
- indicator -> 405 (returned 300 bytes) 
- indicator -> 105 (returned 105 bytes)

You're done when the indicator is less than zero (but not SQL_NO_TOTAL, eg SQL_NULL_DATA) or when the indicator is greater than zero but less than the size of the buffer passed to SQLGetData.

Reply:
Ohhh, I see. There's a whole separate if leg for when it doesn't return SQL_NO_TOTAL on the first go.

I think those two cases could be combined pretty easily since you pretty much have to do the same things in both places. The only real difference is that when it returns SQL_NO_TOTAL, you have can't calculate the final length, so you have pick a size increment to use. Other than that, all the code should be the same I think.

Reply:
It definitely seems possible. I tried to combine them at first but met a little resistance in coding it, will take another stab now that I can conceptualize it a little easier.

Reply:
Should be solved in latest push